larpandoracontent/LArTwoDReco/LArClusterAssociation/CrossGapsExtensionAlgorithm.cc

```diff
@@ -51,72 +51,90 @@ void CrossGapsExtensionAlgorithm::GetListOfCleanClusters(const ClusterList *cons
-    LArPointingClusterList &outerPointingClusterList) const
-{
-    // Convert each input cluster into a pointing cluster
-    LArPointingClusterList pointingClusterList;
+            const float figureOfMerit12(logTheta * (2.f * clusterLength2 - clusterSeparation));
+            const float figureOfMerit21(logTheta * (2.f * clusterLength1 - clusterSeparation));
 
-    for (const Cluster *const pCluster : clusterVector)
-    {
-        try {pointingClusterList.push_back(LArPointingCluster(pCluster));}
-        catch (StatusCodeException &) {}
-    }
+            if (figureOfMerit12 + figureOfMerit21 < std::numeric_limits<float>::epsilon())
+                continue;
+
+            const ClusterAssociation::VertexType vertexType1(closestVertex1.IsInnerVertex() ? ClusterAssociation::INNER : ClusterAssociation::OUTER);
+            const ClusterAssociation::VertexType vertexType2(closestVertex2.IsInnerVertex() ? ClusterAssociation::INNER : ClusterAssociation::OUTER);
 
-    // Identify clusters adjacent to detector gaps
-    this->BuildPointingClusterList(true, pointingClusterList, innerPointingClusterList);
-    this->BuildPointingClusterList(false, pointingClusterList, outerPointingClusterList);
+            (void) clusterAssociationMatrix[pCluster1].insert(ClusterAssociationMap::value_type(pCluster2,
+                ClusterAssociation(vertexType1, vertexType2, ClusterAssociation::STRONG, figureOfMerit12)));
+            (void) clusterAssociationMatrix[pCluster2].insert(ClusterAssociationMap::value_type(pCluster1,
+                ClusterAssociation(vertexType2, vertexType1, ClusterAssociation::STRONG, figureOfMerit21)));
+        }
+    }
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-void CrossGapsExtensionAlgorithm::BuildPointingClusterList(const bool useInner, const LArPointingClusterList &inputPointingClusterList,
-    LArPointingClusterList &outputPointingClusterList) const
+void CrossGapsExtensionAlgorithm::BuildPointingClusterList(const ClusterVector &clusterVector, LArPointingClusterList &pointingClusterList) const
 {
-    for (const LArPointingCluster &pointingCluster : inputPointingClusterList)
+    for (const Cluster *const pCluster : clusterVector)
     {
-        const LArPointingCluster::Vertex &pointingVertex(useInner ? pointingCluster.GetInnerVertex() : pointingCluster.GetOuterVertex());
-        const HitType hitType(LArClusterHelper::GetClusterHitType(pointingCluster.GetCluster()));
-
-        if (LArGeometryHelper::IsInGap(this->GetPandora(), pointingVertex.GetPosition(), hitType, m_maxGapTolerance))
-            outputPointingClusterList.push_back(pointingCluster);
+        try
+        {
+            pointingClusterList.push_back(LArPointingCluster(pCluster));
+        }
+        catch (StatusCodeException &)
```

Comment:
const again.

In the long term, we should consider getting into the habit of checking whether to bother attempting construction of a LArPointingCluster, rather than always trying and adding a blanket catch block, which can hide a number of things and can be so slow if an exception is raised.